### ISSUE TITLE
varnishstat prints null bytes for flag and format

### ISSUE THREAD

**Issue by nigoroll** [varnishcache/varnish-cache#2340]
It does not appear to me that this could be intentional, or is it?

`varnishstat -j | less`

```
  "timestamp": "2017-06-07T14:22:15",
  "MAIN.summs": {
    "description": "stat summ operations",
    "type": "MAIN", "flag": "^@", "format": "^@",
    "value": 0
  },
  "MAIN.uptime": {
    "description": "Child process uptime",
    "type": "MAIN", "flag": "^@", "format": "^@",
    "value": 51
  },
...
```

`varnishstat -x | less`

```
<?xml version="1.0"?>
<varnishstat timestamp="2017-06-07T14:27:56">
        <stat>
                <type>MAIN</type>
                <name>summs</name>
                <value>0</value>
                <flag>^@</flag>
                <format>^@</format>
                <description>stat summ operations</description>
        </stat>
        <stat>
                <type>MAIN</type>
                <name>uptime</name>
                <value>392</value>
                <flag>^@</flag>
                <format>^@</format>
                <description>Child process uptime</description>
        </stat>
...
```

**Comment by nigoroll** [comment]
ping @mbgrydeland ?

**Comment by fgsch** [comment]
revision? I don't think this is intentional. Maybe related to recent work?

**Comment by nigoroll** [comment]
revision: master as always ;) 9f335089c5d2b8b8a45e662dd8f3d8b6f41dec71 (<- the revision for everyone born in December 1st 1971 (not me))

Yes, it seems to be a recent regression:
Though there were no recent changes in the varnishstat code which does not prevent the null character to be output, the actual regression is that the flag/format values are wrong compared with 5.1.2:

```
slink@haggis:/tmp/varnish-cache ((varnish-5.1.2))$ /tmp/512/sbin/varnishd -V
varnishd (varnish-5.1.2 revision 6ece695)
Copyright (c) 2006 Verdens Gang AS
Copyright (c) 2006-2015 Varnish Software AS
slink@haggis:/tmp/varnish-cache ((varnish-5.1.2))$ /tmp/512/sbin/varnishd -a 127.0.0.1:8080 -b 127.0.0.1:80
Debug: Platform: Linux,3.16.0-4-amd64,x86_64,-jnone,-smalloc,-smalloc,-hcritbit
Debug: Child (32379) Started
slink@haggis:/tmp/varnish-cache ((varnish-5.1.2))$ /tmp/512/bin/varnishstat -j |less
slink@haggis:/tmp/varnish-cache ((varnish-5.1.2))$ /tmp/512/bin/varnishstat -j |head
{
  "timestamp": "2017-06-07T17:03:58",
  "MAIN.uptime": {
    "description": "Child process uptime",
    "type": "MAIN", "flag": "c", "format": "d",
    "value": 821
  },
  "MAIN.sess_conn": {
    "description": "Sessions accepted",
    "type": "MAIN", "flag": "c", "format": "i",
```

**Comment by nigoroll** [comment]
OK, `(struct VSC_desc).semantics` and .format got lost. `.semantics` is in the new VSC RST-ish, but not `.format`.

@bsdphk any secret evil plan with this or can I just add the field to the `.vsc` files and `vsctool.py`?

This exemplary patch brings back `.semantics` and sets `.format` to `i`
```
diff --git a/lib/libvarnishapi/vsc.c b/lib/libvarnishapi/vsc.c
index 37334f2..d1bc51b 100644
--- a/lib/libvarnishapi/vsc.c
+++ b/lib/libvarnishapi/vsc.c
@@ -344,17 +344,26 @@ vsc_build_pt_list(struct VSM_data *vd)
                        vdsc = calloc(sizeof *vdsc, 1);
                        AN(vdsc);
 
-#define DOF(n, k)                                              \
+#define DOFL(k)                                                \
                        vt = vjsn_child(vv, k);                 \
                        AN(vt);                                 \
-                       assert(vt->type == VJSN_STRING);        \
-                       vdsc->n = vt->value;
+                       assert(vt->type == VJSN_STRING)
+#define DOF(n,k)                                               \
+                       DOFL(k);                                \
+                       vdsc->n = vt->value
 
                        DOF(name,  "name");
                        DOF(ctype, "ctype");
                        DOF(sdesc, "oneliner");
                        DOF(ldesc, "docs");
+
+                       DOFL("type");
+                       vdsc->semantics = vt->value[0];
+                       // XXX MISSING IN RST-ish / JSON
+                       vdsc->format = 'i';
 #undef DOF
+#undef DOFL
+
                        vdsc->level = &level_info;
                        vt = vjsn_child(vv, "index");
                        AN(vt);
```

**Comment by nigoroll** [comment]
For the most part, this has been addressed in a93c4555ef664ce2835cb43d4cccddd188a27ecf, but some detail needs attention still (diff between the 5.1.2 and master output, editing out the now removed fields `type` and `ident`:
```
$ diff -ub /tmp/varnishstat_512_edited.json /tmp/varnishstat_master.json | egrep -B 2 -A 2 '^-.*flag'
   "MAIN.s_req_hdrbytes": {
     "description": "Request header bytes",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.s_req_bodybytes": {
     "description": "Request body bytes",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.s_resp_hdrbytes": {
     "description": "Response header bytes",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.s_resp_bodybytes": {
     "description": "Response body bytes",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.s_pipe_hdrbytes": {
     "description": "Pipe request header bytes",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.s_pipe_in": {
     "description": "Piped bytes from client",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.s_pipe_out": {
     "description": "Piped bytes to client",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MAIN.bans_persisted_bytes": {
     "description": "Bytes used by the persisted ban lists",
-    "flag": "g", "format": "B",
+    "flag": "g", "format": "i",
     "value": 16
--
   "MAIN.bans_persisted_fragmentation": {
     "description": "Extra bytes in persisted ban lists due to fragmentation",
-    "flag": "g", "format": "B",
+    "flag": "g", "format": "i",
     "value": 0
--
   "MAIN.vsm_free": {
     "description": "Free VSM space",
-    "flag": "g", "format": "B",
-    "value": 973808
+    "flag": "g", "format": "i",
--
   "MAIN.vsm_used": {
     "description": "Used VSM space",
-    "flag": "g", "format": "B",
-    "value": 83960800
+    "flag": "g", "format": "i",
--
   "MAIN.vsm_cooling": {
     "description": "Cooling VSM space",
-    "flag": "g", "format": "B",
+    "flag": "g", "format": "i",
     "value": 0
--
   "MAIN.vsm_overflow": {
     "description": "Overflow VSM space",
-    "flag": "g", "format": "B",
+    "flag": "g", "format": "i",
     "value": 0
--
   "MAIN.vsm_overflowed": {
     "description": "Overflowed VSM space",
-    "flag": "c", "format": "B",
+    "flag": "c", "format": "i",
     "value": 0
--
   "MGT.uptime": {
     "description": "Management process uptime",
-    "flag": "c", "format": "d",
-    "value": 11
+    "flag": "c", "format": "i",
--
   "VBE.boot.default.happy": {
     "description": "Happy health probes",
-    "flag": "b", "format": "b",
+    "flag": "b", "format": "i",
     "value": 0

```

**Comment by hermunn** [comment]
Backport review: I am just going to trust that this is a regression since 4.1. Yell if I am wrong.
